File: C++/ATerm/Gauss.cpp
#include <iostream>
#include <math.h>
#include <windows.h>
#include <iomanip>
#include <string>
#include <quadmath.h>

#define __float128 long double //Arreglo para VSCode, quitar para mayor precision al compilar

using namespace std;


void mostrar(__float128 x[][10], int cx, int lx, string prompt = "Resultado: "){ //Esta es una funcion que muestra matrices

    cout << endl << endl << prompt << endl;
    for (int i = 1; i <= lx; ++i) {
		for (int j = 1; j <= cx; ++j) {

      long double temp = static_cast<long double>(x[i][j]);
      cout.precision(10);
			cout << right << setw(2) << "| "
			<< right << setw(10) << temp << " |";
		}

		cout <<  endl;
    }

    for (int n = 1; n <= cx; ++n) {

		cout << "---------";
		}
		cout << endl;

}
int main(){

  __float128 x[10][10], t[10];
  long double di, nm, temp; // Establece valores como numeros reales, el long double indica que deben ser de alta precision
  int lx, cx, cc; //Los enteros seran usados para contar filas y columnas
  bool found; //Y un booleano para verificar parametros en un futuro
  string stemp;

  cout << "Inserte la cantidad de filas de su matriz x: "; //Empezamos pidiendo parametros de la matriz
	cin >> lx;
	cout << "Inserte la cantidad de columnas de su matriz x: ";
	cin >> cx;

  cx = cx + 1; // Esto añade una columna, que sera la que almacena las constantes

	for (int i = 1; i <= lx; i++) { //Pide que se inserte la matriz

		for (int j = 1; j <= cx; j++) {

      if(j != cx){
        cout << "Inserte la variable en [" << i << "][" << j << "] : ";
        cin >> temp;
        x[i][j] = temp;
      }

      else{
        cout << "Inserte la solucion en la coordenada [" << i << "][" << j << "] : ";
        cin >> temp;
        x[i][j] = temp;
      }

		}
	}

	mostrar(x, cx , lx , "Su matriz es: "); // Muestra el sistema de ecuaciones insertado

    if(lx > cx){ //Verifica si hay mas variables que constantes, si es asi, la matriz tiene infinitas soluciones
        cout << "La matriz tiene soluciones infinitas" << endl;
        system("pause");
        return 0;
    }

    cc = 1; //Columna clave, columna en que debe aparecer un 1 en la matriz en funcion de la fila analizada, para crear una matriz identidad

    for(int i = 1; i <= lx;++i ){

        string paso = string("Paso numero ") + to_string(i);

        if(cc == i && x[i][cc] == 0){ //Trata de arreglar las cosas si la variable que debe ser 1 es 0

            found = false; //Establece en una variable que la fila actual tiene un 0 donde no deberia estar

            if(i < lx){ // Verifica si la fila con el 0 es la ultima, si no, busca filas para intercambiar y areglar el problema
                for( int j = i + 1; j <= lx; ++j){

                    if(x[j][cc] != 0){
                        for(int k = 1; k <= cx; ++k){

                            t[k] = x[i][k];
                            x[i][k] = x[j][k];
                            x[j][k] = t[k];
                        }
                    found = true; //Si logra arreglar el problema, establece en la variable "found" que puede continuar normalmente
                    }
                }
            }

            if(found == false){  //Si no encuentra una fila para intercambiar valida (sin un 0 en la columna clave), o no hay soluciones o hay infinitas
                cout << "Problema encontrado en la fila " << i << endl;
                cout << "La matriz o tiene soluciones infinitas, o no tiene, verificando . . ." << endl;

                if(x[lx][cx] != 0){ //Busca si existe una inconsistencia, si es asi, no existen soluciones
                    cout << "El sistema no tiene soluciones! " << endl;
                }

                else{ //De lo contrario, hay infinitas soluciones
                    cout << "El sistema tiene infinitas soluciones!" << endl;
                }
                system("pause");
                return 0;
            }
        }

        if(cc == i && x[i][cc] != 1){ // Verifico si la coordenada es 1, si no lo es, multiplico toda la fila por su inversa

            di = 1/x[i][cc]; //Inversa del punto de la matriz que debe convertirse en 1

            for(int j = 1; j <= cx; ++j){

                x[i][j] = x[i][j] * di; //Se multiplica toda la fila por esta inversa, de manera que ahora tenemos el 1 que necesitamos


            }

        }

        for (int n = 1;n <= lx; ++n ){ // Ahora reduzco el resto de elementos de la columna a 0 para obtener la forma escalonada completa

            if(n != i){
                nm = x[i][cc] * x[n][cc] * -1; //numero hecho para reducir el resto de elementos de la columna a 0

                for(int c = 1; c <= cx; ++c){
                    x[n][c] = (x[i][c] * nm) + x[n][c]; //nm veces la fila fi + la fila fn
                    if(x[n][c] < 1e-10 && x[n][c] > -1e-10){
                      x[n][c] = 0;
                    }
                }
                mostrar(x, cx, lx , paso);
            }
        }
    cc += 1; //Pasa a la siguiente columna clave, en resumen, hace que los 1 esten en diagonal,
             //creando una matriz identidad con las constantes a la derecha



    mostrar(x, cx, lx , paso);  //Muestra la matriz con la fila i despejada
    cout << endl;
    }


    mostrar(x, cx, lx); //Muestra la matriz resultado

    system("pause");
    return 0;
}

File: PY/DTerm/Gauss.py
import os
import sys
import fractions as fs


def espacio(x=1):
    for _ in range(x):
        print()


def mostrar(x, fx, cx, prompt):
    print(prompt)
    espacio(1)

    for f in range(fx):
        for c in range(cx):
            s = str(x[f][c])
            print("|{:^9}|".format(s), end="")
        print()
        
        for c in range(cx + 1):
            print("".rjust(8, "-"), end="")
        print()

    espacio(1)


def main():

    fx = int(input("Inserte el numero de filas de la matriz: "))
    cx = int(input("Inserte el numero de columnas de la matriz: "))

    x = [[fs.Fraction(int(input("Inserte la coordenada [{}][{}]: ".format(
        y + 1, x + 1))), 1) for x in range(cx)] for y in range(fx)]

    t = [fs.Fraction(int(input(
        "Inserte la solucion de la fila [{}]: ".format(y + 1))), 1) for y in range(fx)]

    for i in range(fx):
        x[i].append(t[i])

    cx = cx + 1

    mostrar(x, fx, cx, "Su matriz es :")


    if fx + 1 < cx:
        print("Filas menores que columnas detectadas")
        print("La matriz tiene infinitas soluciones")
        mostrar(x, fx, cx, "Matriz resultado: ")
        return 0


    cc = 0

    for f in range(fx):  # Va por cada fila de la matriz

        if x[f][cc] == 0 and cc == f:
            found = False

            if f != fx - 1:  # No puedo verificar por filas para intercambiar si ya estoy en la ultima
                for s in range(f, fx):
                    if x[s][cc] != 0:
                        t = x[f]
                        x[f] = x[s]
                        x[s] = t
                        found = True

            if found == False:
                print("La matriz o no tiene soluciones o tiene infinitas soluciones, detectando . . .\n")

                if x[f][cx-1] == 0:
                    print("La matriz tiene infinitos resultados")
                else:
                    print("La matriz no tiene resultados")
                mostrar(x, fx, cx, "Matriz resultante:")
                return 0

        if x[f][cc] != 1 and cc == f:  # Verifica la coordenada que deberia ser 1, sino, la convierte en 1

            di = fs.Fraction(1, x[f][cc])

            for c in range(cx):

                x[f][c] = fs.Fraction(x[f][c] * di, 1)

        for f2 in range(fx):

            if f2 != cc:

                mu = fs.Fraction(-1 * x[f2][cc], 1)

                for c2 in range(cx):
                    x[f2][c2] = fs.Fraction(
                        x[f][c2] * mu, 1) + fs.Fraction(x[f2][c2], 1)

        cc = cc + 1


    mostrar(x, fx, cx, "Resultado:")

    for c in range(cc):
        print("X{} = {}".format(c + 1, x[c][cx - 1]))

    return 0

main()
os.system("pause")
